tutorials/dataframe/df103_NanoAODHiggsAnalysis.C

```diff
@@ -0,0 +1,517 @@
+                                     "Two opposite charged electron and muon pairs");
+   return df_2p2n;
+}
+
+// Reconstruct two Z candidates from four leptons of the same kind
+RVec<RVec<size_t>> reco_zz_to_4l(rvec_f pt, rvec_f eta, rvec_f phi, rvec_f mass, rvec_i charge)
+{
+   RVec<RVec<size_t>> idx(2);
+
+   // Find first lepton pair with invariant mass closest to Z mass
+   const auto z_mass = 91.2;
+   auto idx_cmb = Combinations(pt, 2);
+   auto best_mass = -1;
+   size_t best_i1 = 0;
+   size_t best_i2 = 0;
+   for (size_t i = 0; i < idx_cmb[0].size(); i++) {
+      const auto i1 = idx_cmb[0][i];
+      const auto i2 = idx_cmb[1][i];
+      if (charge[i1] != charge[i2]) {
+         TLorentzVector p1, p2;
+         p1.SetPtEtaPhiM(pt[i1], eta[i1], phi[i1], mass[i1]);
+         p2.SetPtEtaPhiM(pt[i2], eta[i2], phi[i2], mass[i2]);
+         const auto this_mass = (p1 + p2).M();
+         if (std::abs(z_mass - this_mass) < std::abs(z_mass - best_mass)) {
+            best_mass = this_mass;
+            best_i1 = i1;
+            best_i2 = i2;
+         }
+      }
+   }
+   idx[0].push_back(best_i1);
+   idx[0].push_back(best_i2);
+
+   // Reconstruct second Z from remaining lepton pair
+   for (size_t i = 0; i < 4; i++) {
+      if (i != best_i1 && i != best_i2) {
+         idx[1].push_back(i);
+      }
+   }
+
+   // Return indices of the pairs building two Z bosons
+   return idx;
+}
+
+// Compute Z masses from four leptons of the same kind and sort ascending in distance to Z mass
+RVec<float> compute_z_masses_4l(const RVec<RVec<size_t>> &idx, rvec_f pt, rvec_f eta, rvec_f phi, rvec_f mass)
+{
+   RVec<float> z_masses(2);
+   for (size_t i = 0; i < 2; i++) {
+      TLorentzVector p1, p2;
+      const auto i1 = idx[i][0];
+      const auto i2 = idx[i][1];
+      p1.SetPtEtaPhiM(pt[i1], eta[i1], phi[i1], mass[i1]);
+      p2.SetPtEtaPhiM(pt[i2], eta[i2], phi[i2], mass[i2]);
+      z_masses[i] = (p1 + p2).M();
+   }
+   const auto z_mass = 91.2;
+   if (std::abs(z_masses[0] - z_mass) < std::abs(z_masses[1] - z_mass)) {
+      return z_masses;
+   } else {
+      return Reverse(z_masses);
+   }
+}
+
+// Compute mass of Higgs from four leptons of the same kind
+float compute_higgs_mass_4l(const RVec<RVec<size_t>> &idx, rvec_f pt, rvec_f eta, rvec_f phi, rvec_f mass)
+{
+   TLorentzVector p1, p2, p3, p4;
+   const auto i1 = idx[0][0];
+   const auto i2 = idx[0][1];
+   const auto i3 = idx[1][0];
+   const auto i4 = idx[1][1];
+   p1.SetPtEtaPhiM(pt[i1], eta[i1], phi[i1], mass[i1]);
+   p2.SetPtEtaPhiM(pt[i2], eta[i2], phi[i2], mass[i2]);
+   p3.SetPtEtaPhiM(pt[i3], eta[i3], phi[i3], mass[i3]);
+   p4.SetPtEtaPhiM(pt[i4], eta[i4], phi[i4], mass[i4]);
+   return (p1 + p2 + p3 + p4).M();
+}
+
+// Apply selection on reconstructed Z candidates
+RNode filter_z_candidates(RNode df)
+{
+   auto df_z1_cut = df.Filter("Z_mass[0] > 40 && Z_mass[0] < 120", "Mass of first Z candidate in [40, 120]");
```

Comment:
We could template a new operation which is "IsInInterval" which covers all the combinations of collections and scalars and < and <=, all relying on exisitng logical vecops.